dev/tools/localization/gen_l10n.dart

```diff
@@ -302,184 +311,241 @@ bool _isValidGetterAndMethodName(String name) {
+    className = classNameString;
+  }
+
+  /// Sets the reference [Directory] for [l10nDirectory].
+  @visibleForTesting
+  void setL10nDirectory(String arbPathString) {
+    if (arbPathString == null)
+      throw L10nException('arbPathString argument cannot be null');
+    l10nDirectory = _fs.directory(arbPathString);
+    if (!l10nDirectory.existsSync())
+      throw FileSystemException(
+        "The 'arb-dir' directory, $l10nDirectory, does not exist.\n"
+        'Make sure that the correct path was provided.'
+      );
+
+    final FileStat fileStat = l10nDirectory.statSync();
+    if (_isNotReadable(fileStat) || _isNotWritable(fileStat))
+      throw FileSystemException(
+        "The 'arb-dir' directory, $l10nDirectory, doesn't allow reading and writing.\n"
+        'Please ensure that the user has read and write permissions.'
+      );
+  }
 
-    if (FileSystemEntity.isFileSync(entityPath)) {
-      final RegExp arbFilenameRE = RegExp(r'(\w+)\.arb$');
-      if (arbFilenameRE.hasMatch(entityPath)) {
-        final File arbFile = File(entityPath);
-        final Map<String, dynamic> arbContents = json.decode(arbFile.readAsStringSync()) as Map<String, dynamic>;
-        String localeString = arbContents['@@locale'] as String;
+  /// Sets the reference [File] for [templateArbFile].
+  @visibleForTesting
+  void setTemplateArbFile(String templateArbFileName) {
+    if (templateArbFileName == null)
+      throw L10nException('templateArbFileName argument cannot be null');
+    if (l10nDirectory == null)
+      throw L10nException('l10nDirectory cannot be null when setting template arb file');
+
+    templateArbFile = _fs.file(path.join(l10nDirectory.path, templateArbFileName));
+    final String templateArbFileStatModeString = templateArbFile.statSync().modeString();
+    if (templateArbFileStatModeString[0] == '-' && templateArbFileStatModeString[3] == '-')
+      throw FileSystemException(
+        "The 'template-arb-file', $templateArbFile, is not readable.\n"
+        'Please ensure that the user has read permissions.'
+      );
+  }
 
+  /// Sets the reference [File] for the localizations delegate [outputFile].
+  @visibleForTesting
+  void setOutputFile(String outputFileString) {
+    if (outputFileString == null)
+      throw L10nException('outputFileString argument cannot be null');
+    outputFile = _fs.file(path.join(l10nDirectory.path, outputFileString));
+  }
+
+  @visibleForTesting
+  set className(String classNameString) {
+    if (classNameString == null)
+      throw L10nException('classNameString argument cannot be null');
+    if (!_isValidClassName(classNameString))
+      throw L10nException(
+        "The 'output-class', $classNameString, is not a valid Dart class name.\n"
+      );
+    _className = classNameString;
+  }
+
+  /// Scans [l10nDirectory] for arb files and parses them for language and locale
+  /// information.
+  void parseArbFiles() {
+    final List<FileSystemEntity> fileSystemEntityList = l10nDirectory
```

Comment:
Otherwise you're doing an implicit downcast from `FileSystemEntity`

```suggestion
    final List<File> fileSystemEntityList = l10nDirectory
```